README.md
# FFmpeg Video Slideshow Scripts [![Join the chat at https://gitter.im/ffmpeg-video-slideshow-scripts/Lobby](https://badges.gitter.im/ffmpeg-video-slideshow-scripts/Lobby.svg)](https://gitter.im/ffmpeg-video-slideshow-scripts/Lobby?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) [![made-with-bash](https://img.shields.io/badge/Made%20with-Bash-1f425f.svg)](https://www.gnu.org/software/bash/) [![MIT license](https://img.shields.io/badge/License-MIT-blue.svg)](https://lbesson.mit-license.org/)

Shell scripts to create video slideshows from images using [FFmpeg](https://www.ffmpeg.org/).

<img src="https://github.com/tanersener/ffmpeg-video-slideshow-scripts/blob/main/docs/ffmpeg-video-slideshow-scripts-logo-v1.png" width="240">

## Features
- [Transition scripts](transition_scripts) 
- [Advanced scripts](advanced_scripts)


## Requirements

- `ffmpeg 2.8.x` or later

`Advanced Moving Text` script needs `FFmpeg` to be build with `freetype`. If you compile FFmpeg from source you need to provide `--enable-libfreetype` flag on `./configure`. 


## Description

Each script creates a video slideshow using 5 photos inside `photos` folder. Transition scripts 
implements different transition effects and advanced scripts implements more complex animation like 
transitions/transformations.

Output of all scripts is an `h264` encoded `1280x720` px `MPEG-4` video.


## Scene Arrangement
  
Most of the scripts use the following scene arrangement.
```
show photo 1 - 2 seconds
- transition to photo 2 - 1 second
show photo 2 - 2 seconds
- transition to photo 3 - 1 second
show photo 3 - 2 seconds
- transition to photo 4 - 1 second
show photo 4 - 2 seconds
- transition to photo 5 - 1 second
show photo 5 - 2 seconds
=======
Total = 15 seconds
```


## Customization

Scripts in this branch are mostly static and hard to customize. It is possible to change the photos used by replacing files under `photos` folder but other than that there are not any parameters/options defined to customize the slideshow.

If you need to customize scripts please refer to [v2.x](https://github.com/tanersener/ffmpeg-video-slideshow-scripts/tree/v2.x) branch of this repository.


## Versions

- **v1.x branch:** Main focus of this branch is to show how `ffmpeg` filters work and how they can be used to implement a transition or animation. So scripts in this branch are mostly static and hard to customize.  
 
- **v2.x branch:** Scripts are optimized/rearranged in order to be customized. They are hard to understand but easy to customize.


## Updates

Refer to [Changelog](CHANGELOG.md) for updates.


## Tips

- Input images are scaled to fit the screen using `scale=w='if(gte(iw/ih,1280/720),min(iw,1280),-1)':h='if(gte(iw/ih,1280/720),-1,min(ih,720))'` formula. 
As an alternative, cropping input images can be implemented by replacing the scale statement with `crop=min(iw\,1280):min(ih\,720)`.


## Known Issues

You may notice the following warnings when executing the scripts. Below you can find their reasons and possible fixes.

##### 1. 
>[swscaler @ 0x............] deprecated pixel format used, make sure you did set range correctly

This warning is printed because input image streams are decoded with `yuvj444p` pixel format, which is deprecated. 
You can safely ignore it, `ffmpeg` users are not effected from this warning.

##### 2.
>[out_0_0 @ 0x............] 100 buffers queued in out_0_0, something may be wrong.s dup=. drop=. speed=...

>[Parsed_overlay_80 @ 0x............] [framesync @ 0x............] Buffer queue overflow, dropping.

Statements inside `filter_complex` are ordered into logical groups to give a better understanding of how they work. 
In this ordering scheme too many streams/statements wait in the buffer queue, which generates these two warnings. 
If you want to resolve them change the order of statements inside `filter_complex` and use new streams immediately 
after they are created.

##### 3.
>Past duration 0.xyz too large

Currently `push_box` and `box_in` transitions print this warning. `setpts=0.5*PTS` statements used inside the scripts 
cause this. If you know how to remove it safely please submit an issue.


## Useful Links

[How to prevent shake effect for zoompan filter](https://trac.ffmpeg.org/ticket/4298)


## Contributing

Feel free to submit issues or pull requests.


## License
This project is licensed under the [MIT License](https://opensource.org/licenses/MIT) with the following exceptions.

Photos inside `photos` folder are published in the public domain. These photos are:

- [Colosseum](https://www.flickr.com/photos/134331036@N08/35674227104/) by [Klaus Berdiin Jensen](https://www.flickr.com/photos/134331036@N08/)

- [The Great Pyramid & Sphinx Eygpt](https://www.flickr.com/photos/130817154@N04/24211972286/) by [Dave Bright](https://www.flickr.com/photos/130817154@N04/)

- [Leaning Tower of Pisa](https://www.flickr.com/photos/image-catalog/19897194376/) by [Image Catalog](https://www.flickr.com/photos/image-catalog/)

- [Taj Mahal](https://www.flickr.com/photos/149013784@N08/32862668233/) by [juancolado3](https://www.flickr.com/photos/149013784@N08/)

- [chichen itza](https://www.flickr.com/photos/kanvc/15398655930/) by [kan_v_c](https://www.flickr.com/photos/kanvc/)

Snow flake and heart images inside `objects` folder are downloaded from [pngimg.com](http://pngimg.com/download/7553) and [pngimg.com](http://pngimg.com/download/687), both licensed under the [Creative Commons 4.0 BY-NC](https://creativecommons.org/licenses/by-nc/4.0).

Film strip images inside `objects` folder are modified from [Film Strip](https://commons.wikimedia.org/wiki/File:Film_strip.svg) by [Nevit Dilmen](https://commons.wikimedia.org/wiki/User:Nevit) and licensed under [Creative Commons Attribution-Share Alike 3.0 Unported](https://creativecommons.org/licenses/by-sa/3.0/deed.en).

Falling Sky font inside `fonts` folder is licensed under the [SIL Open Font License (OFL)](https://opensource.org/licenses/OFL-1.1).


## See Also

- [FFmpeg Filters](https://ffmpeg.org/ffmpeg-filters.html)
- [FFmpeg Filtering Guide](https://trac.ffmpeg.org/wiki/FilteringGuide)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## - 2023-07-16
- master branch renamed as main

## - 2019-09-08
- Updated the license of scripts with video support

## - 2019-08-24
- Added video slideshow script generator

## - 2019-08-10
- Added audio support for video inputs

## - 2019-07-24
- Added scripts with video input support

## - 2019-06-25
- Enabled travis configuration.

## - 2019-06-08
- Added small gif samples in main README in v2.x branch.

## - 2019-06-04
- Renamed rotate transition as rotate #1 in v2.x branch.

## - 2019-05-26
- Renamed PHOTO_DURATION parameter as IMAGE_DURATION in v2.x branch.
- Moved advanced scripts for push box and sliding bars to transition folder in v2.x branch.

## - 2019-05-25
- Renamed PHOTO_MODE as SCREEN_MODE in v2.x branch.
- Replaced photos folder with media folder in v2.x branch.

## - 2019-05-22
- Updated position formulas in advanced zoom in/out transitions in v2.x branch.

## - 2019-05-21
- Implemented PHOTO_MODE and BACKGROUND_COLOR options in advanced zoom in/out transitions in v2.x branch.

## - 2019-05-19
- Implemented PHOTO_MODE option for rotate transition in v2.x branch.

## - 2019-05-18
- Removed checkpoint phase from box in transition in v2.x branch.

## - 2019-04-12
- Zoom speed parameter introduced for zoom & pan with fade in/out transition #1 in v2.x branch.

## - 2019-03-28
- Added support for transparent input photos on push scripts in v2.x branch.

## - 2019-03-27
- Added support for file names with space character in v2.x branch.

## - 2019-02-16
- Zoom & Pan with Fade In/Out Transition #2 scripts added in v2.x branch.

## - 2019-01-23
- Circular collapse and circular expand transitions added.

## - 2018-12-15
- Tips section added in README.

## - 2018-12-11
- Stack transition implemented in v2.x.

## - 2018-12-10
- Push box moved under advanced scripts in v2.x.
- Direction parameter implemented in v2.x.

## - 2018-12-09
- Photo mode parameter implemented in v2.x.

## - 2018-12-06
- v1.x and v2.x branch documentations updated.

## - 2018-11-30
- v1.x branch created.

## - 2018-08-29
- v1.0 tag created from master.
- Development branch renamed to v2.x.
- Project log added.

## - 2018-08-19
- Spin blur rotation transition implemented.
- Logo Overlay script renamed as Logo Overlay & Zoom.

## - 2018-06-28
- Box In transitions implemented.
- Push Box transitions implemented.

## - 2018-06-27
- Object Move script renamed to Object Animation. New script uses a new snowflake image. Snowflakes are both rotating and falling.
- Logo Overlay updated. New script overlays a beating heart animation instead of static camera image.

## - 2018-06-23
- Wipe In Horizontal with Vertical Panning and Wipe In Vertical with Horizontal Panning scripts updated as gists.
and removed from Advanced Scripts section.
- Zoom In & Pan with Fade In/Out Transition and Zoom Out & Pan with Fade In/Out Transition scripts added in Advanced Scripts.

## - 2018-06-21
- Wipe In Horizontal with Vertical Panning script added in Advanced Scripts.
- Wipe In Vertical with Horizontal Panning script added in Advanced Scripts.

## - 2018-06-16
- Existing bars transition scripts renamed as bars transition #1.
- Bars transition #2 implemented.
- Clock transition added.
- Advanced sliding bars added.

## - 2018-04-07
- Bars transition implemented.

## - 2018-02-06
- Transition and advanced scripts added.

advanced_scripts/object_move.sh
#!/bin/bash
#
# ffmpeg video slideshow script with advanced object move v1 (01.10.2017)
#
# Copyright (c) 2017, Taner Sener (https://github.com/tanersener)
#
# This work is licensed under the terms of the MIT license. For a copy, see <https://opensource.org/licenses/MIT>.
#

START_TIME=$SECONDS

ffmpeg -y \
-loop 1 -i ../photos/1.jpg \
-loop 1 -i ../photos/2.jpg \
-loop 1 -i ../photos/3.jpg \
-loop 1 -i ../photos/4.jpg \
-loop 1 -i ../photos/5.jpg \
-loop 1 -i ../objects/snow-flake.png \
-filter_complex "\
[0:v]setpts=PTS-STARTPTS,scale=w='if(gte(iw/ih,1280/720),min(iw,1280),-1)':h='if(gte(iw/ih,1280/720),-1,min(ih,720))',scale=trunc(iw/2)*2:trunc(ih/2)*2,setsar=sar=1/1,format=rgba,split=2[stream1out1][stream1out2];\
[1:v]setpts=PTS-STARTPTS,scale=w='if(gte(iw/ih,1280/720),min(iw,1280),-1)':h='if(gte(iw/ih,1280/720),-1,min(ih,720))',scale=trunc(iw/2)*2:trunc(ih/2)*2,setsar=sar=1/1,format=rgba,split=2[stream2out1][stream2out2];\
[2:v]setpts=PTS-STARTPTS,scale=w='if(gte(iw/ih,1280/720),min(iw,1280),-1)':h='if(gte(iw/ih,1280/720),-1,min(ih,720))',scale=trunc(iw/2)*2:trunc(ih/2)*2,setsar=sar=1/1,format=rgba,split=2[stream3out1][stream3out2];\
[3:v]setpts=PTS-STARTPTS,scale=w='if(gte(iw/ih,1280/720),min(iw,1280),-1)':h='if(gte(iw/ih,1280/720),-1,min(ih,720))',scale=trunc(iw/2)*2:trunc(ih/2)*2,setsar=sar=1/1,format=rgba,split=2[stream4out1][stream4out2];\
[4:v]setpts=PTS-STARTPTS,scale=w='if(gte(iw/ih,1280/720),min(iw,1280),-1)':h='if(gte(iw/ih,1280/720),-1,min(ih,720))',scale=trunc(iw/2)*2:trunc(ih/2)*2,setsar=sar=1/1,format=rgba,split=2[stream5out1][stream5out2];\
[stream1out1]pad=width=1280:height=720:x=(1280-iw)/2:y=(720-ih)/2:color=#00000000,trim=duration=3,select=lte(n\,90)[stream1overlaid];\
[stream1out2]pad=width=1280:height=720:x=(1280-iw)/2:y=(720-ih)/2:color=#00000000,trim=duration=1,select=lte(n\,30)[stream1ending];\
[stream2out1]pad=width=1280:height=720:x=(1280-iw)/2:y=(720-ih)/2:color=#00000000,trim=duration=2,select=lte(n\,60)[stream2overlaid];\
[stream2out2]pad=width=1280:height=720:x=(1280-iw)/2:y=(720-ih)/2:color=#00000000,trim=duration=1,select=lte(n\,30),split=2[stream2starting][stream2ending];\
[stream3out1]pad=width=1280:height=720:x=(1280-iw)/2:y=(720-ih)/2:color=#00000000,trim=duration=2,select=lte(n\,60)[stream3overlaid];\
[stream3out2]pad=width=1280:height=720:x=(1280-iw)/2:y=(720-ih)/2:color=#00000000,trim=duration=1,select=lte(n\,30),split=2[stream3starting][stream3ending];\
[stream4out1]pad=width=1280:height=720:x=(1280-iw)/2:y=(720-ih)/2:color=#00000000,trim=duration=2,select=lte(n\,60)[stream4overlaid];\
[stream4out2]pad=width=1280:height=720:x=(1280-iw)/2:y=(720-ih)/2:color=#00000000,trim=duration=1,select=lte(n\,30),split=2[stream4starting][stream4ending];\
[stream5out1]pad=width=1280:height=720:x=(1280-iw)/2:y=(720-ih)/2:color=#00000000,trim=duration=2,select=lte(n\,60)[stream5overlaid];\
[stream5out2]pad=width=1280:height=720:x=(1280-iw)/2:y=(720-ih)/2:color=#00000000,trim=duration=1,select=lte(n\,30)[stream5starting];\
[stream2starting][stream1ending]blend=all_expr='A*(if(gte(T,1),1,T/1))+B*(1-(if(gte(T,1),1,T/1)))',select=lte(n\,30)[stream2blended];\
[stream3starting][stream2ending]blend=all_expr='A*(if(gte(T,1),1,T/1))+B*(1-(if(gte(T,1),1,T/1)))',select=lte(n\,30)[stream3blended];\
[stream4starting][stream3ending]blend=all_expr='A*(if(gte(T,1),1,T/1))+B*(1-(if(gte(T,1),1,T/1)))',select=lte(n\,30)[stream4blended];\
[stream5starting][stream4ending]blend=all_expr='A*(if(gte(T,1),1,T/1))+B*(1-(if(gte(T,1),1,T/1)))',select=lte(n\,30)[stream5blended];\
[stream1overlaid][stream2blended][stream2overlaid][stream3blended][stream3overlaid][stream4blended][stream4overlaid][stream5blended][stream5overlaid]concat=n=9:v=1:a=0,format=yuv420p[videowithoutobject];\
[5:v]setpts=PTS-STARTPTS,setsar=sar=1/1,trim=duration=15,select=lte(n\,450),split=5[flake1][flake2][flake3][flake4][flake5];\
[videowithoutobject][flake1]overlay=y='mod(t/2*(1280/3),main_h + 16*overlay_h)-14*overlay_h':x=650:format=rgb,trim=duration=15,select=lte(n\,450)[videowith1flake];\
[videowith1flake][flake2]overlay=y='mod(t/2*(1280/3),main_h + 2*overlay_h)-overlay_h':x=1050:format=rgb,trim=duration=15,select=lte(n\,450)[videowith2flakes];\
[videowith2flakes][flake3]overlay=y='mod(t/2*(1280/3),main_h + 8*overlay_h)-6*overlay_h':x=400:format=rgb,trim=duration=15,select=lte(n\,450)[videowith3flakes];\
[videowith3flakes][flake4]overlay=y='mod(t/2*(1280/3),main_h + 6*overlay_h)-4*overlay_h':x=800:format=rgb,trim=duration=15,select=lte(n\,450)[videowith4flakes];\
[videowith4flakes][flake5]overlay=y='mod(t/2*(1280/3),main_h + 12*overlay_h)-10*overlay_h':x=200:format=rgb,trim=duration=15,select=lte(n\,450)[video]"\
 -map [video] -vsync 2 -async 1 -rc-lookahead 0 -g 0 -profile:v main -level 42 -pix_fmt yuv420p -c:v libx264 -r 30 ../advanced_object_move.mp4

ELAPSED_TIME=$(($SECONDS - $START_TIME))

echo 'Slideshow created in '$ELAPSED_TIME' seconds'